From 3e232c54cc9cf2a9bf9e1aa099c7b790bab5fbc9 Mon Sep 17 00:00:00 2001
From: Christian Brauner <christian.brauner@ubuntu.com>
Date: Wed, 6 Oct 2021 11:56:05 +0200
Subject: [PATCH] os: add separate entries for pure core scheduling kernel
 feature and container support

Allow callers with access to os to tell the difference between kernel
support and container support. This will be used in follow-up patches to
implement core scheduling support for containers even when the shared
library doesn't support it.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>
---
 lxd/daemon.go                      | 6 +++++-
 lxd/instance/drivers/driver_lxc.go | 2 +-
 lxd/sys/os.go                      | 9 +++++----
 3 files changed, 11 insertions(+), 6 deletions(-)

diff --git a/lxd/daemon.go b/lxd/daemon.go
index 6d2d53c1b..8233f41f9 100644
--- a/lxd/daemon.go
+++ b/lxd/daemon.go
@@ -840,9 +840,13 @@ func (d *Daemon) init() error {
 		logger.Info(" - pidfds: no")
 	}
 
-	if canUseCoreScheduling() && d.os.LXCFeatures["core_scheduling"] {
+	if canUseCoreScheduling() {
 		d.os.CoreScheduling = true
 		logger.Info(" - core scheduling: yes")
+
+		if d.os.LXCFeatures["core_scheduling"] {
+			d.os.ContainerCoreScheduling = true
+		}
 	} else {
 		logger.Info(" - core scheduling: no")
 	}
diff --git a/lxd/instance/drivers/driver_lxc.go b/lxd/instance/drivers/driver_lxc.go
index 8dcd96873..0670a89c7 100644
--- a/lxd/instance/drivers/driver_lxc.go
+++ b/lxd/instance/drivers/driver_lxc.go
@@ -739,7 +739,7 @@ func (d *lxc) initLXC(config bool) error {
 		}
 	}
 
-	if d.state.OS.CoreScheduling {
+	if d.state.OS.ContainerCoreScheduling {
 		err = lxcSetConfigItem(cc, "lxc.sched.core", "1")
 		if err != nil {
 			return err
diff --git a/lxd/sys/os.go b/lxd/sys/os.go
index 2688c2ea5..07a91adb2 100644
--- a/lxd/sys/os.go
+++ b/lxd/sys/os.go
@@ -73,6 +73,7 @@ type OS struct {
 
 	// Kernel features
 	CloseRange              bool
+	CoreScheduling          bool
 	NetnsGetifaddrs         bool
 	PidFdSetns              bool
 	SeccompListener         bool
@@ -81,10 +82,10 @@ type OS struct {
 	UeventInjection         bool
 	VFS3Fscaps              bool
 
-	CoreScheduling       bool
-	NativeTerminals      bool
-	PidFds               bool
-	SeccompListenerAddfd bool
+	ContainerCoreScheduling bool
+	NativeTerminals         bool
+	PidFds                  bool
+	SeccompListenerAddfd    bool
 
 	// LXC features
 	LXCFeatures map[string]bool
-- 
2.33.1

